### PR TITLE
feat(oauth2): Enable authorization code grant flow for MuleSoft Connected Apps [CS-5415]

### PR DESCRIPTION
**Pull request by 0xTenable**
Sign-In Button to support `Sign in with MuleSoft` for server-side apps. 

Add a scopes attribute to signin component, changes authType to authorization_code, updates oauth endpoints for the Signin button, and extends styling from the base [Anypoint Button](https://github.com/anypoint-web-components/anypoint-button) component.

This is the same as the Server Side App flow for Google Signin, as illustrated below: 
code flow
See [google's sign-in for server-side-flow](https://developers.google.com/identity/sign-in/web/server-side-flow) for more details.

Here's a [demo page](http://anypoint-signin-test.surge.sh/) with usage of the button for an app in Production today. Use account Oauthtest123 / Oauthtest123 at anypoint.mulesoft.com.
image

**Warning: Breaking Changes**

This PR is to enable usage of the `anypoint-signin button` for server-side apps. Specifically, the button will now support the `authorization_code` grant type OAuth 2.0 flow for the Connected Apps release for the Anypoint Platform. With the Connected Apps release, the Anypoint Platform will become an OAuth 2.0 provider, giving developers the capability to create clients following the OAuth 2.0 protocol to request authorization from users for specific data available on the Anypoint Platform (e.g. manage api alerts, manage apis, read api policies, etc.), a capability that previously wasn't available on the platform.
Note: With these changes, the developer using the `anypoint-signin button` has to handle setting the state of the signin button. 

The button works for the authorization_code flow and simply retrieves the authorization code once the user has granted authorization to the app. The developer has to handle exchanging the code for an access token via the /token endpoint and setting the button's state to signed in afterwards (or, as [Google does for server-side apps](https://developers.google.com/identity/sign-in/web/server-side-flow#step_6_send_the_authorization_code_to_the_server), hide the sign-in button after the user is authorized).

The Anypoint Platform currently does not support client-side flows via the implicit or authorization_code PKCE flows.   

**Added Functionality**

Added the following attributes to the anypoint-signin element: 
  - [x]  **scopes** - the scopes that the application requests authorization from the user for.
  - [x]  **authType** - determines the authorization flow the button uses (e.g. code vs. token).
- [x] Add support for the **authorization_code** flow. This will be the only flow that the anypoint-signin button will initially support for the Connected Apps release. 

**Removed Functionality**

- [x] Removes profile calls from anypoint-signin-aware
- [x] Remove all styling options besides setting width. This is to stay consistent with Anypoint Components on the Anypoint Platform. Styling is based on this [codepen](https://codepen.io/taylordolan/pen/pozdLEx) provided by @taylordolan

### CODE REVIEW

File: anypoint-signin-aware.js
Comment (jarrodek) [inline anypoint-signin-aware.js]: When not implicitly stating what is the type of the property it should be documented with `@type` tag. Here it would be `@type {Array<String>|String}`. However later on you expect a string in `_scopesChanged` function. So we should decide whether only string is acceptable (I think it's OK to make it string only) and then add `type` property `scopes: { type: String }` or we allow to set Array/String. In later case `_scopesChanged` has to be changed to also accept array values.

File: anypoint-signin-aware.js
Comment (jarrodek) [inline anypoint-signin-aware.js:680]: Note my comment for `scope` property. If you decide to support array values then this should be working with array. Array values can be passed as a property (not by the attribute setting), in a templating system for example (React, LitElement, Polymer).

File: anypoint-signin-aware.js
Comment (jarrodek) [inline anypoint-signin-aware.js:682]: And here's a problem with previously set scope and authorization token. 
Should the token be discarded when scope changes?
Situation 1:
Initial scope `profile email`
user: authorized
change scope to: `profile`

Situation 2:
Initial scope `profile`
user: authorized
change scope to: `profile email`

In scenario 1 there's no need to change the button or authentication state as new scope is still valid for the previous one.  No need to do anything. However, in scenario 2 new scope has been added and the user is no longer authorized for this scope. User action is required and button should change it's state to not signed in.
The second situation should inform the application that additional authentication is required (it is not the same as just authentication is required) so the application can perform background auth and/or inform the user about the change.

**Review by jarrodek — CHANGES_REQUESTED** [review]
Initial feedback. Thanks for the PR! It's brilliant

**Comment by jarrodek** [comment]
As for the code flow. The `oauth2-authorization` component does not have any special setting to return code instead of token and will attempt to exchange the code. It will fail. 
For the authorization code flow you have `forceOauthEvents` option which makes the aware component to dispatch `oauth2-token-requested` event instead calling authorize function directly. This event should be handled by the hosting application to initialize OAuth flow with given parameters. It's not well documented but this is how this components handle this situation.

File: anypoint-signin-aware.js
Comment (0xTenable) [reply anypoint-signin-aware.js:682]: I think it's better to inform the user of the change (user always needs to be informed of permission changes). Where would be the best place to set the signed in state to false when the token requests more state than was initially granted?

File: anypoint-signin-aware.js
Comment (0xTenable) [reply anypoint-signin-aware.js:680]: Decided to support string only for now

File: anypoint-signin-aware.js
Comment (0xTenable) [reply anypoint-signin-aware.js]: Set the scope type to String

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js:55]: Please use standard jsdoc comments, right now this file is a bit messy documentation-wise.

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js]: Please remove the copyright/license noise and replace with a very clear statement of purpose for this module.
Also the naming here is inconsistent and thus confusing. The file name does not match the exported object name.

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js:64]: Why repeat AnypointAuth._authType = val;
twice? You always do it.

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js]: Add a TODO to replace this?

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js]: Please unnest into a self-documenting variable name. E.g. useCustomCodeExchange or something.

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js]: This function seems to either return a promise or an undefined value. This is a dangerous practice.

File: anypoint-signin-aware.js
Comment (noambenami) [inline anypoint-signin-aware.js]: This feels like a bad name. Signin-aware is an adjective. It feels like we need a noun here as a name.

**Review by noambenami — DISMISSED** [review]

File: anypoint-signin-aware.js
Comment (jarrodek) [reply anypoint-signin-aware.js]: Agree

File: anypoint-signin-aware.js
Comment (jarrodek) [reply anypoint-signin-aware.js]: It should be marked as `async` and it won't matter

File: anypoint-signin-aware.js
Comment (0xTenable) [reply anypoint-signin-aware.js:64]: Update: Unfortunately it has to be set this way otherwise the values are not set properly. There are tests verifying this functionality.

File: anypoint-signin-aware.js
Comment (0xTenable) [reply anypoint-signin-aware.js]: It doesn't need to return anything anymore since we are no longer fetching the profile

**Comment by jarrodek** [comment]
I have released updated oauth authorisation component so you can test with it.
I can't merge it with stage as it has non-production URIs to the authorisation endpoint. Otherwise it looks good. Some stuff to remove from icons.

File: anypoint-signin-styles.js
Comment (jarrodek) [inline anypoint-signin-styles.js]: This is the same as in the AnypointButton class

**Comment by 0xTenable** [comment]
Comments from Pawel: 
> ok, when you are setting a boolean property to "true" in constructor, there's no way to override this using attributes when declaring the element in a HTML template. compatibility and noink is set to true in the PR. You either remove it or introduce new attribute that disables them when connectedCallback is called (this is where attributes are translated into properties).
disabled,  elevation and toggles already have default values and you don't need to declare them.
> also compatibility implies noink
> (https://github.com/anypoint-web-components/anypoint-button/blob/master/src/AnypointButton.js#L182)

File: anypoint-signin-styles.js
Comment (jarrodek) [inline anypoint-signin-styles.js]: Use css variables for that. In ACR I won't be using any of this fonts.

File: anypoint-signin-styles.js
Comment (jarrodek) [inline anypoint-signin-styles.js]: Recently I started using `1rem` as a default value for body font weight. It's better solution than hard coding pixel value.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js:176]: Is this `div` really required here? I doesn't seems to have any purpose

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: You have removed themed styling so this should be removed as well.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: This is not necessary as it does nothing. undefined is default value but this implies that the button is not disabled.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: Elevation is computed here: https://github.com/anypoint-web-components/anypoint-button/blob/master/src/AnypointButtonBase.js#L103
There's no point of setting it here. When compatibility is set the styling renders the button without the shadow.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js:334]: It is acceptable to set a default value here to "high" however this might now be the most important button on the application screen. Emphasis reflects button importance on the page. By default it should be low. Otherwise it makes the UI busy and the user may be confused about that should be done. 
It's just to consider. I don't care about it as much in this case.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: This is implied with `compatibility` set.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: There is no dark theme anymore?

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: This is default value inherited from ButtonStateMixin.

File: anypoint-signin.js
Comment (jarrodek) [inline anypoint-signin.js]: I don't like this property here to be set to `true`. To set it to false it would require programmatic access to the button after it gets upgraded (meaning defined and initialized). I understand that this should be the default value in this case. Please, introduce a new property (like `material` or whatever) and when this property is set to true then in sets `compatibility` to `false`.

File: mulesoft-icons.js
Comment (jarrodek) [inline mulesoft-icons.js]: Can we remove this? There's nothing here.

**Review by jarrodek — CHANGES_REQUESTED** [review]

File: anypoint-signin.js
Comment (0xTenable) [reply anypoint-signin.js]: Doesn't seem like it -- I've just removed this. I think you have one place in your demo that mentions darkTheme in Anypoint Button too -- might be good to remove https://github.com/anypoint-web-components/anypoint-button/blob/b6222c7126793ffe7b77c6df8b67f9e60d8e9862/demo/index.js#L86

File: anypoint-signin-styles.js
Comment (0xTenable) [reply anypoint-signin-styles.js]: What's the css variable you have for font-family? var(--arc-font-family)?

File: anypoint-signin.js
Comment (0xTenable) [reply anypoint-signin.js:176]: It's so I can add the buttonText class to the text and signIn / signOut class for testing

File: anypoint-signin-styles.js
Comment (jarrodek) [reply anypoint-signin-styles.js]: You can introduce new one. It doesn't have to be ARC theme's variable

File: anypoint-signin.js
Comment (jarrodek) [reply anypoint-signin.js:176]: But there's no style definition for it.

Comment (jarrodek) [reply anypoint-signin.js:176]: unless I missed something

Comment (0xTenable) [reply anypoint-signin.js:176]: .buttonText is defined in anypoint-signin-styles.js right here :D https://github.com/anypoint-web-components/anypoint-signin/pull/2/files#diff-4192739dbb3588b7b30324de15750817R17

**Review by jarrodek — APPROVED** [review]

**Pull request merged**
